hazelcast/util.py

```diff
@@ -345,3 +346,112 @@ def create_git_info():
+            return cmp(self.obj, other.obj) > 0
+
+        def __eq__(self, other):
+            return cmp(self.obj, other.obj) == 0
+
+        def __le__(self, other):
+            return cmp(self.obj, other.obj) <= 0
+
+        def __ge__(self, other):
+            return cmp(self.obj, other.obj) >= 0
+
+        def __ne__(self, other):
+            return cmp(self.obj, other.obj) != 0
+    return K
+
+
+ITERATION_TYPE = enum(KEY=0, VALUE=1, ENTRY=2)
+
+
+def get_sorted_query_result_set(result_list_future, paging_predicate):
+    """
+    :param result_list_future, a future that returns a list of (K,V) pairs to sort and slice based on paging_predicate
+    attributes.
+    :param paging_predicate
+    :return: list of sorted query results
+    """
+    result_list = result_list_future.result()
+    if len(result_list) == 0:
+        return []
+
+    comparator = paging_predicate.get_comparator()
+    iteration_type = paging_predicate.get_iteration_type()
+
+    result_list.sort(key=_get_comparison_key(comparator, iteration_type))
+
+    nearest_anchor_entry = paging_predicate.get_nearest_anchor_entry()
+    nearest_page = nearest_anchor_entry[0]  # first element in anchor entry pair is index of nearest page.
+    page = paging_predicate.get_page()
+    page_size = paging_predicate.get_page_size()
+    list_size = len(result_list)
+
+    begin = page_size * (page - nearest_page - 1)
+    if begin > list_size:
+        return []
+    end = min(begin + page_size, list_size)
+
+    _set_anchor(result_list, paging_predicate, nearest_page)
+
+    sorted_query_result_set = result_list[begin:end]
+
+    if iteration_type == ITERATION_TYPE.ENTRY:
+        return sorted_query_result_set[:]
+    elif iteration_type == ITERATION_TYPE.KEY:
+        return [x[0] for x in sorted_query_result_set]
+    elif iteration_type == ITERATION_TYPE.VALUE:
+        return [x[1] for x in sorted_query_result_set]
+
+
+def _set_anchor(result_list, paging_predicate, nearest_page):
+    """
+    :param result_list, a list of (K,V) pairs to sort and slice based on paging_predicate
+    :param paging_predicate
+    :param nearest_page is the index of the page that the last anchor belongs to
+    """
+    list_size = len(result_list)
+    assert list_size > 0
+    page = paging_predicate.get_page()
+    page_size = paging_predicate.get_page_size()
+    for i in range(page_size, list_size, page_size):
+        if nearest_page < page:
+            anchor = result_list[i-1]
+            nearest_page += 1
+            paging_predicate.set_anchor(nearest_page, anchor)
+
+
+def _get_comparison_key(comparator, iteration_type):
+    """
+    :param comparator: Comparator object if specified in paging predicate, or None.
+    :param iteration_type
+    :return: key function to be used in sorting
+    """
+    if comparator is not None:
+        return cmp_to_key(comparator.compare)
+    else:
+        if iteration_type == ITERATION_TYPE.KEY:
+            return lambda entry: entry[0]
```

Comment:
Good point. I thought in that case Python would compare based on their hash functions but it might not work that way.

The [Paging Predicate section in the reference manual](https://docs.hazelcast.org/docs/latest/manual/html-single/index.html#filtering-with-paging-predicates) says:
> If a comparator is not specified for PagingPredicate, but you want to get a collection of keys or values page by page, this collection must be an instance of Comparable (i.e., it must implement java.lang.Comparable). Otherwise, the java.lang.IllegalArgument exception is thrown.

So perhaps I can add to the documentation that if the user does not specify a  custom comparator, their keys or values should be of a Python-comparable type, that is it should implement the __lt__ method.
Would this solve the issue?
Or should I somehow check whether I can compare the keys or values and raise an error if not?

Reply:
Yes, I guess that's the best we can do